proof/crefine/RISCV64/Schedule_C.thy

```diff
@@ -593,35 +593,957 @@ lemma length_scRefills_bounded:
+               apply (clarsimp simp: maxPeriodUs_def timer_defs unat_max_word)
+              apply ceqv
+             apply (rule_tac val="from_bool (rTime head < maxReleaseTime)"
+                         and xf'=ret__int_'
+                          in ccorres_symb_exec_r_known_rv[where R=\<top> and R'=UNIV])
+                apply (rule conseqPre, vcg)
+                apply (clarsimp simp: crefill_relation_def maxReleaseTime_def maxBound_word
+                               split: if_splits)
+               apply ceqv
+              apply (clarsimp simp: when_def)
+              apply (rule ccorres_cond[where R=\<top>])
+                apply fastforce
+               apply (rule ccorres_pre_getObject_sc, rename_tac sc)
+               apply (rule ccorres_rhs_assoc)+
+               apply (rule ccorres_move_c_guard_sc)
+               apply (rule_tac val="refill_C (rTime head + scPeriod sc) usage"
+                           and xf'=used___struct_refill_C_'
+                           and R="obj_at' (\<lambda>sc'. scPeriod sc' = scPeriod sc) scPtr"
+                            in ccorres_symb_exec_r_known_rv[where R'=UNIV])
+                  apply (rule conseqPre, vcg)
+                  apply normalise_obj_at'
+                  apply (frule (1) obj_at_cslift_sc)
+                  apply (clarsimp simp: crefill_relation_def typ_heap_simps csched_context_relation_def)
+                 apply ceqv
+                apply (rule ccorres_rhs_assoc2)
+                apply (rule ccorres_split_nothrow_novcg[where r'=dc and xf'=xfdc])
+                    apply (clarsimp simp: setRefillHd_def updateRefillHd_def)
+                    apply (rule_tac P'="sc_at' scPtr and invs'"
+                                 in updateSchedContext_ccorres_lemma3
+                                     [where P="\<lambda>sc. scRefillHead sc < length (scRefills sc)"])
+                      apply vcg
+                     apply fastforce
+                    apply clarsimp
+                    apply (rename_tac sc sc')
+                    apply (frule (1) sc_ko_at_valid_objs_valid_sc'[OF _ invs_valid_objs'])
+                    apply (frule rf_sr_refill_buffer_relation)
+                    apply (frule_tac n="scRefillHead sc" in h_t_valid_refill, fastforce+)
+                    apply (intro conjI)
+                       apply (clarsimp simp: typ_heap_simps)
+                      apply (rule disjCI2)
+                      apply (rule_tac n="length (scRefills sc)" in array_assertion_shrink_right)
+                       apply (fastforce intro: sc_at_array_assertion')
+                      apply (clarsimp simp: typ_heap_simps csched_context_relation_def)
+                     apply (clarsimp simp: typ_heap_simps' sc_ptr_to_crefill_ptr_def
+                                           csched_context_relation_def)
+                    apply (clarsimp simp: typ_heap_simps csched_context_relation_def)
+                    apply normalise_obj_at'
```

Comment:
unexpected to see `normalise_obj_at'` so distant from the subgoal-solving statement, anything interesting going on herer?

Reply:
Maybe some of these `normalise_obj_at'` can be removed or moved elsewhere. Maybe I don't understand what you mean by it being distant from the subgoal-solving statement. Were you expecting it to be further up or further down?

Reply:
Well, normally I see it used, and then the subgoal is discharged immediately, or on the next step. This one goes on for a bit until it's actually discharged, making me wonder if something is going the long way around.